urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/lp/solver/BasicCohoSolver.java

```diff
@@ -0,0 +1,865 @@
+			}
+			int removeCol = currBasis.basis().V(evictCol).intValue();// get the column number of the i^th of current basis.
+			//The new basis is dual feasible and primal unknown
+			return currBasis.replace(removeCol,newCol,LPBasis.BasisStatus.UNKNOWN,LPBasis.BasisStatus.FEASIBLE);
+		}catch(SingularMatrixException e){
+			//e.printStackTrace();
+			throw new LPError("CohoSolver.pivot(): This should never happen. The basis should be well-conditon.",e);
+		}
+	}
+
+	/*
+	 * This function is used in LPProject to speedup computation. 
+	 * cohoFeasibleBasis is the optimal basis of previous optimal direction, thus feasible for cohoLP 
+	 * and infeasible for cohoDualLP.
+	 * evict is the index of infeasible variable of cohoFeasibleBasis.
+	 * evict is not the index of cols to remove, it is the index of cohoFeasibleBasis to remove. 
+	 */	
+	//public static int counter1 = 0, counter2=0, counter3=0;
+	public	CohoSolverResult opt(Matrix c, LPBasis cohoFeasibleBasis, int evict){
+//		System.out.println("Direction:"+c);
+//		coho.lp.project.LPProject.oneHopLP++;
+		init(c);
+		LPBasis optBasis;
+		CohoSolverResult result = null;
+		//Try to find optimal basis from cohoFeasibleBasis within one hop
+		try{
+			optBasis= findOptInOneHop(cohoFeasibleBasis, evict);
+		}catch(UnboundedLPError e){//unbounded problem
+			if(isDual)
+				result = new CohoSolverResult(LPResult.ResultStatus.UNBOUNDED,new CohoDouble(Double.MAX_VALUE),(LPBasis)null,(Matrix)null);
+			else
+				result = new CohoSolverResult(LPResult.ResultStatus.INFEASIBLE,null,(LPBasis)null,(Matrix)null);
+//			System.out.println("Optimal Basis:"+result.optBasis().toString());
+			return result;
+		}
+		//coho.debug.STAT.timers[3] += System.nanoTime()-time;
+		//If failed, solve it from scrach
+		if(optBasis==null){
+			//coho.lp.project.LPProject.failLP++;
+			result = opt();
+//			System.out.println("Optimal Basis:"+result.optBasis().toString());
+			return result;
+		}		
+        //find a feasible basis
+		if(optBasis.primalStatus()!=LPBasis.BasisStatus.FEASIBLE){
+			//coho.lp.project.LPProject.feasibleLP++;
+			result = opt(optBasis);
+//			System.out.println("Optimal Basis:"+result.optBasis().toString());
+			return result;
+		}
+		if(isDual){			
+			result = cohoDualResult(optBasis); 
+		}else{
+			result = cohoResult(optBasis);
+		}
+		//coho.lp.project.LPProject.sucessLP++;
+//		System.out.println("Optimal Basis:"+result.optBasis().toString());
+		return result;
+	}
+	
+	/*
+	 * Try to find the optimal basis from startBasis within one hop.
+	 * return null if not a feasible basis found. 
+	 * return the optimal basis if one optimal basis is found.
+	 * return the feasible basis if one feasible basis is found.
+	 *  
+	 */
+//	public static int sameCounter = 0,diffCounter1=0;
+//	public static int same = 0, diff = 0, fail = 0;
+	public static int oneHopMethod = 3;
+	protected LPBasis findOptInOneHop(LPBasis startBasis,int evict)throws UnboundedLPError{
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 23. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.